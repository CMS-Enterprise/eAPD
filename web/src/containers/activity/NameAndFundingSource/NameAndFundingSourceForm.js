import { ChoiceList, TextField } from '@cmsgov/design-system-core';
import PropTypes from 'prop-types';
import React, { Fragment, useCallback } from 'react';

const NameAndFundingSourceForm = ({
  item: { fundingSource, key, name },
  handleChange
}) => {
  const changeName = useCallback(({ target: { value } }) => {
    handleChange(key, { name: value });
  });

  const changeFundingSource = useCallback(({ target: { value } }) => {
    handleChange(key, { fundingSource: value }, true);
  });

  return (
    <Fragment>
      <TextField
        label="Activity name"
        name="activity-name"
        value={name}
        onChange={changeName}
      />
      <ChoiceList
        label="Program type"
        name="program-type"
        choices={[
          { checked: fundingSource === 'HIT', label: 'HIT', value: 'HIT' },
          { checked: fundingSource === 'HIE', label: 'HIE', value: 'HIE' },
          { checked: fundingSource === 'MMIS', label: 'MMIS', value: 'MMIS' }
        ]}
        onChange={changeFundingSource}
      />
    </Fragment>
  );
};

NameAndFundingSourceForm.propTypes = {
  handleChange: PropTypes.func.isRequired,
  item: PropTypes.shape({
    fundingSource: PropTypes.string.isRequired,
    key: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired
};

export default NameAndFundingSourceForm;
