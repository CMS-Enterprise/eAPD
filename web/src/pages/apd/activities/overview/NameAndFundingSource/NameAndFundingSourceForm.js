import { ChoiceList, TextField } from '@cmsgov/design-system';
import PropTypes from 'prop-types';
import React, { Fragment, useEffect } from 'react';
import { useForm, Controller } from 'react-hook-form';
import { joiResolver } from '@hookform/resolvers/joi';
import { connect } from 'react-redux';

import nameFundingSourceSchema from '@cms-eapd/common/schemas/nameAndFundingSource';
import {
  setActivityName,
  setActivityFundingSource
} from '../../../../../redux/actions/editActivity';

const NameAndFundingSourceForm = ({
  index,
  item: { fundingSource, name },
  setFundingSource,
  setName,
  adminCheck
}) => {
  NameAndFundingSourceForm.displayName = 'NameAndFundingSourceForm';

  const {
    control,
    trigger,
    formState: { errors }
  } = useForm({
    defaultValues: {
      name: name,
      fundingSource: fundingSource
    },
    resolver: joiResolver(nameFundingSourceSchema)
  });

  useEffect(() => {
    if (adminCheck) {
      trigger(['name', 'fundingSource']);
    }
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  const choices = ['HIT', 'HIE', 'MMIS'].map(choice => ({
    checked: fundingSource === choice,
    label: choice,
    value: choice
  }));

  return (
    <Fragment>
      <Controller
        name="name"
        control={control}
        render={({ field: { onChange, ...props } }) => (
          <TextField
            {...props}
            label="Activity name"
            onChange={({ target: { value } }) => {
              onChange(value);
              setName(index, value);

              if (adminCheck) {
                trigger();
              }
            }}
            className="remove-clearfix"
            errorMessage={errors?.name?.message}
            errorPlacement="bottom"
          />
        )}
      />
      <Controller
        name="fundingSource"
        control={control}
        render={({ field: { onChange, ...props } }) => (
          <ChoiceList
            {...props}
            choices={choices}
            label="Program type"
            labelClassName="ds-u-margin-bottom--1"
            onChange={({ target: { value } }) => {
              onChange(value);
              setFundingSource(index, value);

              if (adminCheck) {
                trigger();
              }
            }}
            type="radio"
            errorMessage={errors?.fundingSource?.message}
            errorPlacement="bottom"
          />
        )}
      />
    </Fragment>
  );
};

NameAndFundingSourceForm.propTypes = {
  index: PropTypes.number.isRequired,
  item: PropTypes.shape({
    fundingSource: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired,
  setFundingSource: PropTypes.func.isRequired,
  setName: PropTypes.func.isRequired,
  adminCheck: PropTypes.bool.isRequired
};

const mapStateToProps = state => {
  return {
    adminCheck: state.apd.adminCheck.enabled
  };
};

const mapDispatchToProps = {
  setFundingSource: setActivityFundingSource,
  setName: setActivityName
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(NameAndFundingSourceForm);

export {
  NameAndFundingSourceForm as plain,
  mapStateToProps,
  mapDispatchToProps
};
