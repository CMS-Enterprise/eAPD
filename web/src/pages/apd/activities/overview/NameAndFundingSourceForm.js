import { ChoiceList, TextField } from '@cmsgov/design-system';
import PropTypes from 'prop-types';
import React, { Fragment, useCallback } from 'react';
import { connect } from 'react-redux';

import {
  setActivityName,
  setActivityFundingSource
} from '../../../../actions/editActivity';

const NameAndFundingSourceForm = ({
  index,
  item: { fundingSource, name },
  setFundingSource,
  setName
}) => {
  const changeName = useCallback(
    ({ target: { value } }) => {
      setName(index, value);
    },
    [index, setName]
  );

  const changeFundingSource = useCallback(
    ({ target: { value } }) => {
      setFundingSource(index, value);
    },
    [index, setFundingSource]
  );

  const choices = ['HIT', 'HIE', 'MMIS'].map(choice => ({
    checked: fundingSource === choice,
    label: choice,
    value: choice
  }));

  return (
    <Fragment>
      <TextField
        label="Activity name"
        name="activity-name"
        value={name}
        onChange={changeName}
        className="remove-clearfix"
      />
      <ChoiceList
        choices={choices}
        label="Program type"
        labelClassName="ds-u-margin-bottom--1"
        name="program-type"
        onChange={changeFundingSource}
        type="radio"
      />
    </Fragment>
  );
};

NameAndFundingSourceForm.propTypes = {
  index: PropTypes.number.isRequired,
  item: PropTypes.shape({
    fundingSource: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired,
  setFundingSource: PropTypes.func.isRequired,
  setName: PropTypes.func.isRequired
};

const mapDispatchToProps = {
  setFundingSource: setActivityFundingSource,
  setName: setActivityName
};

export default connect(null, mapDispatchToProps)(NameAndFundingSourceForm);

export { NameAndFundingSourceForm as plain, mapDispatchToProps };
