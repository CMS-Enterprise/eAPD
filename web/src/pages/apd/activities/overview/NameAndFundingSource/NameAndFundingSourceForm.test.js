import React from 'react';
import {
  renderWithConnection,
  act,
  screen,
  waitFor
} from 'apd-testing-library';

import { plain as NameAndFundingSourceForm } from './NameAndFundingSourceForm';

const defaultProps = {
  index: 1,
  item: {
    fundingSource: 'Uncle Scrooge',
    key: 'key 1',
    name: 'Buying bikes for Huey, Dewey, and Louie'
  },
  setFundingSource: jest.fn(),
  setName: jest.fn()
};

const setup = async (props = {}) => {
  const utils = await act(async () => 
    renderWithConnection(<NameAndFundingSourceForm {...defaultProps} {...props} />)
  );
  await waitFor(() => screen.findByText(/Activity Name/i));
  return utils;
};

describe('the NameAndFundingSourceForm component', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test('it renders correctly', async () => {
    await setup();
    expect(
      screen.getByLabelText(/Activity name/i)
    ).toHaveValue(defaultProps.item.name)
  });
})