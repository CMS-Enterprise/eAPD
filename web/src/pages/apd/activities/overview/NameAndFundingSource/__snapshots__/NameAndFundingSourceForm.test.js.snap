// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the activity name and funding source component renders correctly 1`] = `
<Fragment>
  <Component
    control={
      Object {
        "_defaultValues": Object {
          "fundingSource": "Uncle Scrooge",
          "name": "Buying bikes for Huey, Dewey, and Louie",
        },
        "_executeSchema": [Function],
        "_fields": Object {},
        "_formState": Object {
          "dirtyFields": Object {},
          "errors": Object {},
          "isDirty": false,
          "isSubmitSuccessful": false,
          "isSubmitted": false,
          "isSubmitting": false,
          "isValid": false,
          "isValidating": false,
          "submitCount": 0,
          "touchedFields": Object {},
        },
        "_formValues": Object {
          "fundingSource": "Uncle Scrooge",
          "name": "Buying bikes for Huey, Dewey, and Louie",
        },
        "_getDirty": [Function],
        "_getFieldArray": [Function],
        "_getWatch": [Function],
        "_names": Object {
          "array": Set {},
          "mount": Set {},
          "unMount": Set {},
          "watch": Set {},
        },
        "_options": Object {
          "defaultValues": Object {
            "fundingSource": "Uncle Scrooge",
            "name": "Buying bikes for Huey, Dewey, and Louie",
          },
          "mode": "onBlur",
          "reValidateMode": "onBlur",
          "resolver": [Function],
          "shouldFocusError": true,
        },
        "_proxyFormState": Object {
          "dirtyFields": false,
          "errors": "all",
          "isDirty": false,
          "isValid": false,
          "isValidating": false,
          "touchedFields": false,
        },
        "_removeUnmounted": [Function],
        "_stateFlags": Object {
          "action": false,
          "mount": false,
          "watch": false,
        },
        "_subjects": Object {
          "array": Object {
            "next": [Function],
            "observers": Array [],
            "subscribe": [Function],
            "unsubscribe": [Function],
          },
          "state": Object {
            "next": [Function],
            "observers": Array [],
            "subscribe": [Function],
            "unsubscribe": [Function],
          },
          "watch": Object {
            "next": [Function],
            "observers": Array [],
            "subscribe": [Function],
            "unsubscribe": [Function],
          },
        },
        "_updateFieldArray": [Function],
        "_updateValid": [Function],
        "getFieldState": [Function],
        "register": [Function],
        "unregister": [Function],
      }
    }
    name="name"
    render={[Function]}
  />
  <Component
    control={
      Object {
        "_defaultValues": Object {
          "fundingSource": "Uncle Scrooge",
          "name": "Buying bikes for Huey, Dewey, and Louie",
        },
        "_executeSchema": [Function],
        "_fields": Object {},
        "_formState": Object {
          "dirtyFields": Object {},
          "errors": Object {},
          "isDirty": false,
          "isSubmitSuccessful": false,
          "isSubmitted": false,
          "isSubmitting": false,
          "isValid": false,
          "isValidating": false,
          "submitCount": 0,
          "touchedFields": Object {},
        },
        "_formValues": Object {
          "fundingSource": "Uncle Scrooge",
          "name": "Buying bikes for Huey, Dewey, and Louie",
        },
        "_getDirty": [Function],
        "_getFieldArray": [Function],
        "_getWatch": [Function],
        "_names": Object {
          "array": Set {},
          "mount": Set {},
          "unMount": Set {},
          "watch": Set {},
        },
        "_options": Object {
          "defaultValues": Object {
            "fundingSource": "Uncle Scrooge",
            "name": "Buying bikes for Huey, Dewey, and Louie",
          },
          "mode": "onBlur",
          "reValidateMode": "onBlur",
          "resolver": [Function],
          "shouldFocusError": true,
        },
        "_proxyFormState": Object {
          "dirtyFields": false,
          "errors": "all",
          "isDirty": false,
          "isValid": false,
          "isValidating": false,
          "touchedFields": false,
        },
        "_removeUnmounted": [Function],
        "_stateFlags": Object {
          "action": false,
          "mount": false,
          "watch": false,
        },
        "_subjects": Object {
          "array": Object {
            "next": [Function],
            "observers": Array [],
            "subscribe": [Function],
            "unsubscribe": [Function],
          },
          "state": Object {
            "next": [Function],
            "observers": Array [],
            "subscribe": [Function],
            "unsubscribe": [Function],
          },
          "watch": Object {
            "next": [Function],
            "observers": Array [],
            "subscribe": [Function],
            "unsubscribe": [Function],
          },
        },
        "_updateFieldArray": [Function],
        "_updateValid": [Function],
        "getFieldState": [Function],
        "register": [Function],
        "unregister": [Function],
      }
    }
    name="fundingSource"
    render={[Function]}
  />
</Fragment>
`;
